File: src/element/type.ts
import { Pass } from "../pass/pass";
import { Statement } from "../statement/statement";
import { Element, ElementPriority, ElementType } from "./element";

export abstract class Type<T> implements Element {
    public getPriority(): ElementPriority {
        return ElementPriority.NORMAL;
    }

    constructor(public id: string, public plural: string, public type: Function) {}

    public getElementType(): ElementType {
        return ElementType.TYPE;
    }

    public abstract visit(pass: Pass, statement: Statement): boolean;
    public abstract initialize();
    public abstract canParse(): boolean;
    public abstract fromStringRepresentation(input: string): T;
    public abstract toStringRepresentation(input: T): string;
}

File: common/src/main/kotlin/me/honkling/skriptkt/common/syntax/Type.kt
package me.honkling.skriptkt.common.syntax

import me.honkling.skriptkt.common.parser.Parser
import me.honkling.skriptkt.common.statement.Statement

abstract class Type<T>(val id: String, val plural: String) : SyntaxElement() {
    override fun accept(statement: Statement) {}
    override fun initialize(parser: Parser, statement: Statement): Boolean = true

    abstract fun canParse(): Boolean
    abstract fun fromStringRepresentation(input: String): T
    abstract fun toStringRepresentation(input: T): String
}
